File: scripts/archive/feather-probe.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Fcntl qw(SEEK_SET SEEK_END);
use JSON::PP;

# Read-only format probe. No Arrow decoding, schema validation or whole-file reads.
# The native reader will be selected after the runtime and sample format are confirmed.
my $path = shift @ARGV;
die "Usage: perl scripts/archive/feather-probe.pl FILE.feather\n" unless defined $path && !@ARGV;
open my $fh, '<:raw', $path or die "Cannot open input file: $!\n";
my @info = stat($fh);
die "Input must be a regular file\n" unless -f $fh;
my $size = $info[7];
my $bytes_read = 0;
sub read_at {
    my ($offset, $length) = @_;
    seek($fh, $offset, SEEK_SET) or die "Cannot seek input: $!\n";
    my $buffer = '';
    my $read = read($fh, $buffer, $length);
    die "Cannot read input: $!\n" unless defined $read;
    $bytes_read += $read;
    return $buffer;
}
my $head = read_at(0, $size < 6 ? $size : 6);
my $format = 'unknown';
my $envelope = JSON::PP::false;
my $footer_bytes;
if (substr($head, 0, 6) eq 'ARROW1' && $size >= 18) {
    $format = 'feather_v2_or_arrow_ipc';
    my $tail = read_at($size - 10, 10);
    $footer_bytes = unpack('V', substr($tail, 0, 4));
    $envelope = JSON::PP::true if substr($tail, 4, 6) eq 'ARROW1' && $footer_bytes > 0 && $footer_bytes <= $size - 18;
} elsif (substr($head, 0, 4) eq 'FEA1' && $size >= 12) {
    $format = 'feather_v1';
    my $tail = read_at($size - 8, 8);
    $footer_bytes = unpack('V', substr($tail, 0, 4));
    $envelope = JSON::PP::true if substr($tail, 4, 4) eq 'FEA1' && $footer_bytes > 0 && $footer_bytes <= $size - 12;
}
close $fh or die "Cannot close input: $!\n";
my %modules;
for my $module ('Glib::Object::Introspection', 'FFI::Platypus') {
    eval "require $module";
    $modules{$module} = $@ ? JSON::PP::false : JSON::PP::true;
}
print JSON::PP->new->canonical->pretty->encode({
    file_bytes => "$size", format => $format, envelope_consistent => $envelope,
    footer_bytes => $footer_bytes, bytes_read => $bytes_read,
    schema_validated => JSON::PP::false, compression => 'uninspected',
    perl_version => sprintf('%vd', $^V), available_modules => \%modules,
    note => 'Signature check only. A compatible Arrow reader must still validate schema, compression and record batches.',
});
exit($envelope ? 0 : 2);

File: packages/archive/src/feather-probe.ts
import { open } from 'node:fs/promises';

/**
 * Read-only Feather/Arrow envelope probe. Signature check only — it reads a
 * handful of bytes from the head and tail and never touches the data body, so
 * a multi-gigabyte file is inspected as cheaply as a tiny one. It intentionally
 * does not validate the Arrow schema, decode any record batch, or inspect
 * compression: choosing and running a real reader is a separate, later stage.
 *
 * This is the TypeScript successor to `scripts/archive/feather-probe.pl`; the
 * runtime for the archive/backtesting work was fixed to TypeScript/Node.
 */

const V2_MAGIC = Buffer.from('ARROW1', 'ascii'); // 6 bytes
const V1_MAGIC = Buffer.from('FEA1', 'ascii'); // 4 bytes

// Arrow IPC "File" format (Feather V2): 6-byte magic + 2 pad bytes at the head,
// and a <int32 footer length LE><6-byte magic> trailer. The smallest possible
// well-formed file is head(8) + a non-empty footer + trailer(10) => >= 18.
const V2_TAIL = 10;
const V2_MIN_SIZE = 18;

// Legacy Feather V1: 4-byte magic at head, and a <int32 length LE><4-byte magic>
// trailer. Smallest well-formed file is head(4) + footer + trailer(8) => >= 12.
const V1_TAIL = 8;
const V1_MIN_SIZE = 12;

export type FeatherFormat = 'feather_v2_or_arrow_ipc' | 'feather_v1' | 'unknown';

export interface FeatherProbeResult {
  /** File size in bytes. */
  fileBytes: number;
  /** Detected container format from magic bytes alone. */
  format: FeatherFormat;
  /**
   * Whether head magic, trailing magic and the declared footer length are
   * mutually consistent. `false` means truncated, corrupt, or not this format.
   */
  envelopeConsistent: boolean;
  /** Declared footer length in bytes, or null when it could not be read. */
  footerBytes: number | null;
  /** How many bytes the probe actually read from disk. */
  bytesRead: number;
  /** Always false: the probe never validates the Arrow schema. */
  schemaValidated: false;
  /** Always 'uninspected': the probe never decodes the body. */
  compression: 'uninspected';
  note: string;
}

/**
 * Minimal positional-read source so the pure probe logic can be unit-tested
 * against in-memory buffers as well as real files.
 */
export interface RandomAccessSource {
  readonly size: number;
  /** Read `length` bytes starting at `offset`. May return fewer at EOF. */
  readAt(offset: number, length: number): Promise<Buffer>;
}

const NOTE =
  'Signature check only. A compatible Arrow reader must still validate schema, ' +
  'compression and record batches before any data is trusted.';

function unknown(fileBytes: number, bytesRead: number): FeatherProbeResult {
  return {
    fileBytes,
    format: 'unknown',
    envelopeConsistent: false,
    footerBytes: null,
    bytesRead,
    schemaValidated: false,
    compression: 'uninspected',
    note: NOTE,
  };
}

/** Probe an already-opened random-access source. Pure aside from the reads. */
export async function probeFeatherSource(source: RandomAccessSource): Promise<FeatherProbeResult> {
  const size = source.size;
  let bytesRead = 0;

  const headLen = Math.min(size, 6);
  const head = await source.readAt(0, headLen);
  bytesRead += head.length;

  if (head.subarray(0, 6).equals(V2_MAGIC) && size >= V2_MIN_SIZE) {
    const tail = await source.readAt(size - V2_TAIL, V2_TAIL);
    bytesRead += tail.length;
    const footerBytes = tail.length >= 4 ? tail.readUInt32LE(0) : null;
    const trailingMagicOk = tail.subarray(4, 10).equals(V2_MAGIC);
    const envelopeConsistent =
      trailingMagicOk &&
      footerBytes !== null &&
      footerBytes > 0 &&
      footerBytes <= size - V2_MIN_SIZE;
    return {
      fileBytes: size,
      format: 'feather_v2_or_arrow_ipc',
      envelopeConsistent,
      footerBytes,
      bytesRead,
      schemaValidated: false,
      compression: 'uninspected',
      note: NOTE,
    };
  }

  if (head.subarray(0, 4).equals(V1_MAGIC) && size >= V1_MIN_SIZE) {
    const tail = await source.readAt(size - V1_TAIL, V1_TAIL);
    bytesRead += tail.length;
    const footerBytes = tail.length >= 4 ? tail.readUInt32LE(0) : null;
    const trailingMagicOk = tail.subarray(4, 8).equals(V1_MAGIC);
    const envelopeConsistent =
      trailingMagicOk &&
      footerBytes !== null &&
      footerBytes > 0 &&
      footerBytes <= size - V1_MIN_SIZE;
    return {
      fileBytes: size,
      format: 'feather_v1',
      envelopeConsistent,
      footerBytes,
      bytesRead,
      schemaValidated: false,
      compression: 'uninspected',
      note: NOTE,
    };
  }

  return unknown(size, bytesRead);
}

/** Wrap an in-memory buffer as a random-access source. */
export function bufferSource(buffer: Buffer): RandomAccessSource {
  return {
    size: buffer.length,
    async readAt(offset, length) {
      return buffer.subarray(offset, offset + length);
    },
  };
}

/** Probe a file on disk, reading only its head and tail. */
export async function probeFeatherFile(path: string): Promise<FeatherProbeResult> {
  const handle = await open(path, 'r');
  try {
    const stat = await handle.stat();
    if (!stat.isFile()) {
      throw new Error(`Input must be a regular file: ${path}`);
    }
    const source: RandomAccessSource = {
      size: stat.size,
      async readAt(offset, length) {
        const buffer = Buffer.alloc(length);
        const { bytesRead } = await handle.read(buffer, 0, length, offset);
        return buffer.subarray(0, bytesRead);
      },
    };
    return await probeFeatherSource(source);
  } finally {
    await handle.close();
  }
}
